licenses/license.go

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) 2015-2024 MinIO, Inc.
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// Package license provides functions for verifying MinIO licenses.
+//
+// A simple way to do that is:
+//
+//	func init() {
+//		l, err := license.Verify()
+//		if err != nil {
+//			fmt.Fprintln(os.Stderr, "Error:", err)
+//			os.Exit(1)
+//		}
+//		if time.Now().After(l.ExpiresAt) {
+//			fmt.Fprintf(os.Stderr, "Warning: license expired on %v! Renew and restart immediately to avoid outage.\n", l.ExpiresAt)
+//		}
+//		if time.Now().After(l.NotAfter) {
+//			fmt.Fprintln(os.Stderr, "Error: license has expired. Terminating process...")
+//			os.Exit(1)
+//		}
+//		go func() {
+//			<-time.NewTimer(time.Until(l.ExpiresAt)).C
+//			fmt.Fprintf(os.Stderr, "Warning: license expired on %v! Process will be terminated soon. Renew and restart immediately to avoid outage.\n", l.ExpiresAt)
+//
+//			<-time.NewTimer(time.Until(l.NotAfter)).C
+//			fmt.Fprintln(os.Stderr, "Error: license has expired. Terminating process...")
+//			os.Exit(1)
+//		}()
+//	}
+//
+// For a smooth developer experience, consider enabling license verification only for release builds.
+// This can be achieved via build tags. For example, create a license.go file containing the init
+// function from above and use the build tag "//go:build !dev". When building the binary using
+// "go build -tags dev", license verification is disabled. Otherwise, a license is required.
+package license
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/lestrrat-go/jwx/jwa"
+	"github.com/lestrrat-go/jwx/jwk"
+	"github.com/lestrrat-go/jwx/jwt"
+
+	"time"
+)
+
+const publicKey = `-----BEGIN PUBLIC KEY-----
+MHYwEAYHKoZIzj0CAQYFK4EEACIDYgAEaK31xujr6/rZ7ZfXZh3SlwovjC+X8wGq
+qkltaKyTLRENd4w3IRktYYCRgzpDLPn/nrf7snV/ERO5qcI7fkEES34IVEr+2Uff
+JkO2PfyyAYEO/5dBlPh1Undu9WQl6J7B
+-----END PUBLIC KEY-----`
```

Comment:
We may need some mechanism to get new key in case it changes on SUBNET side.

Reply:
You would only ever want to change a key if it is compromised. If that is the case, the new key should be added here and this one should be removed (after a grace period). It seems reasonable to expect people to upgrade within this timeframe.

Dynamically fetching the key allows someone intercepting the network traffic to replace it and craft their own custom licenses.

Reply:
We have to provide fallbacks when that happens so we can have two keys at a time. The one that we always use , and the one that can be injected freshly into the implementation.

Reply:
> All products doing license verification must immediately be released with the new key (as existing versions will fail verification of newly generated licenses)

All products must be released with a single dep change: Upgraded `minio/pkg`. This can be done as a regular release or (if not possible at the moment) from the latest stable with a backport. Also, there is "immediate" action required since if the key is compromised, any existing binary / container image can still be used. This more on us then the users/customers.

> Any existing customer whose license has expired (and grace period also over) must upgrade to the new version after getting their license extended/renewed (as the new license will be generated with new key)

Technically yes. However, in cases where this is not possible we could make an exception and provide a license signed with the compromised key. 

> If we don't want to enforce upgrade in these situations, one option could be to allow the customer to override the public key by setting the new one in an env or standard file path.

This is even worse then allowing the key to be fetched from the network. If you can provide your own public key for verifying the binary, then you just create a private/public key pair, issue an arbitrary license with the private key and provide the public key you just generated.

Reply:
Let's assume our private signing key `K1` gets compromised. This means someone can issue arbitrary licenses that any exisiting release (up to this day) considers valid. Would this be different if we had e.g. a 2nd public key in the binary? No, but customers would not be required to update when they renew their license.

If `K1` gets compromised, we have to make new releases of all binaries that require a license since existing binaries always accept licenses issued by `K1`. We have no proper way of telling these binaries that they should no longer accept licenses issued by `K1`. We have two options now:
 - Build a binary that only accepts licenses issued by our new key `K2`. This requires that we give customers a new license.
 - Build a binary that accepts licenses issued by `K1` or `K2`. In this case, the customer does not have to update it's license.

Including two public keys (`P1` and `P2`) in the binary **only** allows customers to keep their current version of the binary but use a new license - issued by `K2`.

***

> Existing binary cannot be used by new customers because their license would have been created with the new key and hence would fail verification by existing binary.

This is not clear to me. If `K1` is compromised and a customer wants to use a binary that only accepts licenses issued by `K1` we still can issue a license using `K1` for this customer. Basically:
If customer uses release before `K1` got compromised, issue license with `K1`. Otherwise issue license using `K2`.

Reply:
Let's assume our private signing key `K1` gets compromised.
Now, someone can issue arbitrary licenses that any existing
binary considers valid. Hence, we switch to `K2` and start
issue licenses with it from now on.

Now, there are:
 - Users running existing binaries with licenses issued by `K1`.
   This is fine.
 - Users running existing binaries with licenses issued by `K2`.
   This might be a problem.
 - Users running new binaries with licenses issued by `K1`.
   This might be a problem.
 - Users running new binaries with licenses issued by `K2`.
   This is fine.

When a user has a license issued by `K1` and upgrades its binary
to a version requiring `K2` then the user has to update its license.
This is always the case.

When a user has a license isssued by `K2` and tries to run a
an existing binary then `P2` (public key) must be already in
the binary or it will not work. Otherwise, the user is forced to
upgrade.

Hence, putting a 2nd public key (`P2`) into the binaries now
allows users to update their license without upgrading their
binaries. However, this requires that `K2` is stored separatley
from `K1` (e.g. offline).

An alternative approach would be to issue the license based
on the user's binary version. If the binary still requires a
license issued by `K1` we can issue one. Nothing prevents
that.

In any case, users that upgrade will require a new license eventually.

Reply:
>Existing binary cannot be used by new customers because their license would have been created with the new key and hence would fail verification by existing binary.

What I'm trying to say is that a customer does not have to upgrade immediately. We only start issuing new licenses when releases are ready - not as soon as we notice that the key got compromised. Once new releases are ready, new customers will use them most probably. (Special cases can be taken care of). There is no point in changing the key as long as no releases that can verify it are available.